File: 2023-02/python/part2.py
import math
import re

CUBE_REGEX = re.compile(r"(\d+) (\w+)")


def fewest_cubes_prod(game_string):
    min_cubes = dict(red=0, green=0, blue=0)
    for amount, color in CUBE_REGEX.findall(game_string):
        min_cubes[color] = max(int(amount), min_cubes[color])
    return math.prod(min_cubes.values())


if __name__ == "__main__":
    print(sum((fewest_cubes_prod(line) for line in open("../input.txt", "r").readlines())))

File: 2023-02/rust/part2.rs
use regex::Regex;
use std::cmp::max;
use std::collections::HashMap;
use std::fs::read_to_string;

fn run(filename: &str) -> i32 {
    let cube_regex = Regex::new(r"(\d+) (\w+)").unwrap();

    read_to_string(filename)
        .unwrap()
        .lines()
        .map(|line| {
            let mut min_cubes = HashMap::with_capacity(3);
            for c in cube_regex.captures_iter(line) {
                let amount = c[1].parse::<i32>().unwrap();
                min_cubes
                    .entry(c[2].to_string())
                    .and_modify(|e| *e = max(amount, *e))
                    .or_insert(amount);
            }
            min_cubes.values().product::<i32>()
        })
        .sum()
}

fn main() {
    let result = run("../input.txt");
    println!("{result}");
}
